### PR TITLE
review for smart-home-application

### PR DESCRIPTION
**Pull request by kyorri**
Added the application for review.
I also added a sample file for what the XML file should look like.

### CODE REVIEW

File: smart-home-application/Device.cpp [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+void Device::setStatus(bool status) {
+    this->status = status;
+};
+
+bool Device::getStatus() {
+    return this->status;
+};
+
+std::vector<Sensor> Device::getSensors() {
```
Comment (CristianIonutIonescu) [inline smart-home-application/Device.cpp]: here you're returning a copy of the sensors that are contained by your device. Read about returning references vs values (a bit similar to passing them as arguments) and how that affects your logic.

Long story short, you have a silent bug in your code waiting to trigger

File: smart-home-application/App.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+#include <vector>
+
+class App {
+    public:
+        App(std::string);
+        void save();
+        void load();
+        void addSensor(Device&, std::string, double);
+        void addDevice(Room&, std::string, bool);
+        void newRoom(std::string);
```
Comment (CristianIonutIonescu) [inline smart-home-application/App.h]: even though Google Style guide recommend using CamelCase notation, I'm ok with this notation as well, but make sure you stay consistent for the rest of the course

File: smart-home-application/README.md [hunk 3]
```diff
@@ -0,0 +1,14 @@
+# Smart Home Application
+
+> Here i will explain briefly the chosen data structures to store the information that will be read/written from/to XML files, and explain why I have chosen this approach.
+
+1. I composed the Device class using a bunch of Sensor objects (using std::vector), so that every device will take the sensors' readings.
```
Comment (CristianIonutIonescu) [inline smart-home-application/README.md:5]: Can't the sensors exist independently of the devices?
Eg. I have a very big room that requires more than what an AC unit could sustain therefore I need to install 2. But I don't need 2 sensors to read the room temperature but one to whom both AC units are connected.

I'm not saying it's wrong the way you do it (didn't specify exact requirements) but take it as some food for thought

**Review by CristianIonutIonescu — CHANGES_REQUESTED** [review]
Good work on the code!! As you wrote, there are a few things that still need to be done but there are some other things that need to be done as well:
- to add xml read/write and add/delete objects for final part
- namespaces and scoping
- you have a silent bug when returning values of std::vector, you'll see it when you start implementing your todo
- in your code there's no need for rule of 5, but could you try to force it somewhere to implement it as well?
- [CRITICAL] Read about differences between passing/returning by value  & reference and update your code to respect that. Read about const variables and functions and use it in your code (there are some places where they should be used)

LE: please remove the previous homework code from PR as it makes things harder to review

**Comment by kyorri** [comment]
Resolved changes for review.

File: smart-home-application/Device.cpp [hunk 4]
```diff
@@ -0,0 +1,63 @@
+#include "Device.h"
+
+#include <iostream>
+#include <vector>
+#include <algorithm>
+
+namespace smart_home_app {
+Device::Device(std::string device_name, bool status) : device_name_(device_name), status_(status) {};
+
+Device::~Device() {};
```
Comment (CristianIonutIonescu) [inline smart-home-application/Device.cpp:10]: you can make this one default as well

File: smart-home-application/Device.cpp [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+
+void Device::SetStatus(bool status) {
+    this->status_ = status;
+};
+
+bool Device::GetStatus() {
+    return this->status_;
+};
+
+void Device::AddSensor(Sensor sensor) {
```
Comment (CristianIonutIonescu) [inline smart-home-application/Device.cpp:45]: although is correct, you should pass this as a const reference to improve performance. read a bit more about const and references

this comment applies to other methods as well

**Review by CristianIonutIonescu — APPROVED** [review]
Nicely done, there are a few other things that need to be solved, but the overall homework is complete. 

PS: I didn't comment on all the things that should be changed, leaving some for the following homeworks :) when we'll discuss about them.

**Pull request closed without merging**